src/main/java/com/streamr/client/StreamrClient.java

```diff
@@ -156,23 +170,50 @@ public WebSocketClient getWebsocket() {
         return websocket;
     }
 
+    private void reconnect() {
+        initWebsocket();
+        websocket.connect();
+        waitForState(State.Connected);
+        if (getState() == State.Connected) {
+            StreamrClient.this.subs.forEach(StreamrClient.this::resubscribe);
+        }
+    }
+
     /**
      * Connects the websocket. Blocks until connected, or throws if the connection times out.
      */
     public void connect() throws ConnectionTimeoutException {
-        log.info("Connecting to " + options.getWebsocketApiUrl() + "...");
+        connect(true);
+    }
+
+    private void connect(boolean firstTrial) throws ConnectionTimeoutException {
+        if (state == State.Connected) {
+            log.warn("Trying to connect when already connected to " + options.getWebsocketApiUrl());
+            return;
+        } else if (state == State.Connecting) {
+            log.warn("Trying to connect when already connecting to " + options.getWebsocketApiUrl());
+            return;
```

Comment:
Should have `waitForState(State.Connected);` in this branch? Otherwise the method is not blocking until 
`state == State.Connected`, contrary to what is stated:

> Connects the websocket. Blocks until connected, or throws if the connection times out.